File: src/Flare.Mobile/Services/ServerUrlValidator.cs
namespace Flare.Mobile.Services;

public static class ServerUrlValidator
{
    public static bool TryNormalize(string? input, bool allowHttpLoopback, out string normalized, out string error)
    {
        normalized = string.Empty;
        error = string.Empty;
        if (!Uri.TryCreate(input?.Trim(), UriKind.Absolute, out var uri) || string.IsNullOrWhiteSpace(uri.Host))
        {
            error = "Enter a complete Flare server URL.";
            return false;
        }
        var loopbackHttp = allowHttpLoopback && uri.Scheme == Uri.UriSchemeHttp && uri.IsLoopback;
        if (uri.Scheme != Uri.UriSchemeHttps && !loopbackHttp)
        {
            error = "Flare requires an HTTPS server.";
            return false;
        }
        if (!string.IsNullOrEmpty(uri.UserInfo) || !string.IsNullOrEmpty(uri.Query) || !string.IsNullOrEmpty(uri.Fragment))
        {
            error = "Use the server base URL without credentials, a query, or a fragment.";
            return false;
        }
        normalized = uri.GetLeftPart(UriPartial.Authority).TrimEnd('/');
        return true;
    }
}

File: src/Flare.Mobile/lib/core/api/server_url_validator.dart
final class ServerUrlValidator {
  const ServerUrlValidator._();

  static String normalize(String input, {bool allowHttpLoopback = false}) {
    final trimmed = input.trim();
    final candidate = trimmed.contains('://') ? trimmed : 'https://$trimmed';
    final uri = Uri.tryParse(candidate);
    if (uri == null ||
        uri.host.isEmpty ||
        uri.userInfo.isNotEmpty ||
        uri.hasQuery ||
        uri.hasFragment) {
      throw const FormatException('Enter a valid Flare server URL.');
    }
    final loopback =
        uri.host == 'localhost' ||
        uri.host == '127.0.0.1' ||
        uri.host == '10.0.2.2';
    if (uri.scheme != 'https' &&
        !(allowHttpLoopback && loopback && uri.scheme == 'http')) {
      throw const FormatException('Flare requires an HTTPS server URL.');
    }
    if (uri.path != '' && uri.path != '/') {
      throw const FormatException('Use the server origin without an API path.');
    }
    return uri
        .replace(path: '', query: null, fragment: null)
        .toString()
        .replaceFirst(RegExp(r'/+$'), '');
  }
}
